README.md
# Audero Flashing Text #
[Audero Flashing Text](https://github.com/AurelioDeRosa/Audero-Flashing-Text) is a cross-browser jQuery plugin that creates the effect of a flashing, randomly-placed and randomly-sized text inside a given HTML element (typically a `<div>`).

## Demo ##
A live demo is available [here](http://htmlpreview.github.io/?https://github.com/AurelioDeRosa/Audero-Flashing-Text/blob/master/demo/index.html).

## Requirements ##
Being a jQuery plugin, the only requirement is [jQuery](http://www.jquery.com) starting from version **1.2**.

## Compatibility ##
It has been tested and works on all the major browsers, including Internet Explorer 6 and later.

## Installation ##
You can install Audero Flashing Text by using [Bower](http://bower.io).

    bower install audero-flashing-text

Alternatively, you have to manually download it.

Remember to include the script **after** the [jQuery](http://www.jquery.com) library in the header of your web page:

    <script src="path/to/jquery.js"></script>
    <script src="path/to/jquery.auderoFlashingText.js"></script>

## Usage ##
Using this plugin is very simple. Just call the `auderoFlashingText()` method on the element you want to apply the effect.
For example, let that you have the following code:

    <div id="box">
       <p>First</p>
       <p>Second</p>
       <p>Third</p>
    </div>

A basic call to the plugin, that will rotate the text of the three paragraphs is:

    <script>
       $(window).load(function () {
          $("#box").auderoFlashingText();
       });
    </script>

**Note**: It's suggested to start the animation after the window is fully loaded, that is as a callback of the <code>load</code> event of the <code>window</code> object, because there is no point in running the effect before the user is able to see the complete page.

### Disable the effect ###
After you initialize the plugin, you have the chance to disable it at any time. To disable the effect, you have to call the `auderoFlashingText()` method passing the string `disable`, as you can see in the next example:

    <script>
       $("#box").auderoFlashingText('disable');
    </script>

### Enable the effect ###
Once you disabled the plugin you may want to enable it again. To do this, you have to call the `auderoFlashingText()` method passing the string `enable` as shown in the next snippet:

    <script>
       $("#box").auderoFlashingText('enable');
    </script>

### Toggle the effect ###
In some cases you may want to toggle the effect, for example based on a user action. Let's say that you want to enable/disable the animation when a user clicks a button. To achieve this goal, you have to call the `auderoFlashingText()` method passing the string `toggle`.
So, let that you have the following HTML code:

    <div id="box">
       <p>First</p>
       <p>Second</p>
       <p>Third</p>
    </div>
    <button id="toggle-button">Toggle animation</button>

You can bind the `click` event on the `<button>` to toggle the animation as shown in the following snippet:

    $("#toggle-button").click(function() {
       $("#box").auderoFlashingText('toggle');
    });

### Destroy the effect ###
Under some circumstances you may want to stop the animation and clean all the resources. To perform this task, you can call the `auderoFlashingText()` method passing the string `destroy`. To see it in action, take a look at the following snippet:

    <script>
       $("#box").auderoFlashingText('destroy');
    </script>

## Options ##
Audero Flashing Text has several options you can set during the call to the `auderoFlashingText()` method that allow you to customize the effect to fit your needs. The options are:

* `strings` (`array`. Default: `null`, but if it isn't specified the text of the children of the elements is used): A set of strings to show.
* isEnabled: true (`boolean`. Default: `true`): The state of the animation. If the value is `false`, the animation will stop keeping its current options
* `fadeIn` (`numeric`. Default: `300`): The time in milliseconds the element appears by fading to opaque.
* `duration` (`numeric`. Default: `500`): The time of milliseconds the element stays visible.
* `fadeOut` (`numeric`. Default: `300`): The time of milliseconds the element disappears by fading to transparent.
* `selection` (`string`. Default: `random`): The order of selection of the strings. The possible values are: `random`, `ascending`, `descending`.
* `repeat` (`number`. Default: `-1`): The number of times to repeat the animation. -1 means unlimited.
* `pause` (`number`. Default: `0`): The time (milliseconds) between animations. Set to `random` to have a random time.
* `fontMinSize` (`number`. Default: `7`): The minimum font's size of the strings.
* `fontMaxSize` (`number`. Default: `28`): The maximum font's size of the strings.
* `fontUnit` (`string`. Default: `px`): The unit to use for the font's size. The value accepted are all those accepted by the browsers, for example: `px`, `em`, `rem` and so on.

### Override default values ###
[Audero Flashing Text](https://github.com/AurelioDeRosa/Audero-Flashing-Text) has been developed following the current best practices in developing plugins for jQuery. Therefore, it exposes the previously cited options through the `defaults` object, allowing you to override the properties' default value. This is very useful if you usually use the same values to run the animation. In fact, changing the default values, you don't need to specify them again when you initialize the effect.
For example, let that you have the following code:

    <script>
       $(window).load(function() {
          $("#box-1").auderoFlashingText({
             repeat: 5,
             pause: 500,
             duration: 1000
          });
          $("#box-2").auderoFlashingText({
             repeat: 5,
             pause: 500,
             duration: 2000
          });
       });
    </script>

Overriding the default values you can turn it into the following:

    <script>
       $(window).load(function() {
          $.fn.auderoFlashingText.defaults.repeat = 5;
          $.fn.auderoFlashingText.defaults.pause = 500;
          $("#box-1").auderoFlashingText({
             duration: 1000
          });
          $("#box-2").auderoFlashingText({
             duration: 2000
          });
       });
    </script>

## Advanced Examples ##
### Example 1 ###
This example demonstrates how you can customize the animation passing options to the initialization method. In this example almost all the properties seen in the previous section are used:

    <script>
       $(window).load(function () {
          $("#box").auderoFlashingText({
             fadeIn: 500,
             duration: 800,
             fadeOut: 500,
             pause: 500,
             selection: "descending",
             fontUnit: "em",
             fontMinSize: 0.5,
             fontMaxSize: 3,
          });
       });
    </script>

### Example 2 ###
This second example is slightly more complex then the previous. In fact, it shows you how to select the strings to show using the `children()` and `map()` methods of jQuery.
Let that you have this code:

    <div id="box-2">
       <ul>
          <li>Lorem</li>
          <li>Ipsum</li>
          <li>Dolor</li>
          <li>Sit</li>
          <li>Amet</li>
       </ul>
    </div>

You can run the script in this way:

    <script>
       $(window).load(function () {
          $("#box-2").auderoFlashingText({
             strings: $("#box-2 > ul").children("li").map(function() {
                return $(this).text();
             }),
             fadeOut: 1500,
             selection: "ascending"
          });
       });
    </script>

## License ##
[Audero Flashing Text](https://github.com/AurelioDeRosa/Audero-Flashing-Text) is dual licensed under [MIT](http://www.opensource.org/licenses/MIT) and [GPL-3.0](http://opensource.org/licenses/GPL-3.0)

## Authors ##
[Aurelio De Rosa](http://www.audero.it) ([@AurelioDeRosa](https://twitter.com/AurelioDeRosa))

HISTORY.md
# Audero Flashing Text Changes History #

## 2.0.1 (2014-01-12) ##

* Added bower.json

## 2.0.0 (2013-07-22) ##

* Complete refactor
* Added new properties to allow for more customizations: `isEnabled`, `repeat`, `pause`, `fontMinSize`, `fontMaxSize`, and `fontUnit`
* Removed public methods `isRunning()` and `start()`
* Renamed `stop()` method to `destroy()`
* Added new methods: `enable()`, `disable()`, `toggle()`
* Exposed publicly the properties to allow changing default values
* Improved documentation
* Added more examples in the demo page

## 1.0.0 (2013-04-11) ##

* First release

src/jquery.auderoFlashingText.js
/*
 * Audero Flashing Text is a cross-browser jQuery plugin that creates
 * the effect of a flashing, randomly-placed and randomly-sized text inside
 * a given HTML element (typically a <div>).
 *
 * @author  Aurelio De Rosa <aurelioderosa@gmail.com>
 * @version 2.0.1
 * @link    https://github.com/AurelioDeRosa/Audero-Flashing-Text
 * @license Dual licensed under MIT (http://www.opensource.org/licenses/MIT)
 * and GPL-3.0 (http://opensource.org/licenses/GPL-3.0)
 */
(function ($) {
   // Private properties and methods
   var dataAttributeName = "audero-flashing-text";
   var animateText = function ($element, index) {
      // Save the last used index. In this way, if the user disables the animation,
      // once it run again, the plugin know the last used string
      $.extend($element.data(dataAttributeName), {lastIndex: index});
      var options = $element.data(dataAttributeName);
      var $text = $("<span>")
         .text(options.strings[index])
         .css({
            position: "absolute",
            display: "none",
            fontSize: (Math.random() * options.fontMaxSize + options.fontMinSize) + options.fontUnit
         })
         .appendTo($element)
         .fadeIn(options.fadeIn)
         .animate({opacity: 1}, options.duration) // Simulate delay
         .fadeOut(options.fadeOut, function () {
            // Remove the current element
            $(this).remove();
            var nextIndex;
            if (options.selection === "ascending") {
               nextIndex = (index + 1) % options.strings.length;
            } else if (options.selection === "descending") {
               nextIndex = (index === 0) ? options.strings.length : index - 1;
            } else {
               nextIndex = Math.floor(Math.random() * options.strings.length);
            }

            if (options.repeat > 0) {
               options.repeat--;
            }
            // Generate a random pause
            var time = (options.pause === "random") ? Math.random() * 3000 : options.pause;

            $element.data(dataAttributeName, options);
            // If the animation is enabled test if it should run again
            if (options.isEnabled === true) {
               // If the repetition limit is not reached, the animation method will run again
               if (options.repeat !== 0) {
                  setTimeout(function () { animateText($element, nextIndex); }, time);
               } else {
                  methods.destroy.apply([$element]);
               }
            }
         });
      // Set the position so the element will fit the box's size
      var posX = Math.floor(Math.random() * ($element.width() - $text.outerWidth()));
      var posY = Math.floor(Math.random() * ($element.height() - $text.outerHeight()));
      // Set the position of the text
      $text.css({
         left: posX + "px",
         top: posY + "px"
      });
   };
   var isRunning = function ($element) {
      var options = $element.data(dataAttributeName);
      return (options !== undefined && options.isEnabled === true);
   };

   // Public methods
   var methods = {
      init: function (options) {
         if (typeof options === "undefined" || options === null) {
            options = {};
         }
         if (options.strings == null && this.find(':first').length !== this.length) {
            $.error("To run jQuery.auderoFlashingText if you don't specify the texts to show, each element must have at least a child");
            return;
         }
         // By default the order is ascending
         var index = 0;
         if (options.selection === "descending") {
            index = options.strings.length - 1;
         } else if (options.selection === "random") {
            index = Math.floor(Math.random() * options.strings.length);
         }
         options = $.extend(true, {}, $.fn.auderoFlashingText.defaults, options);

         // Check if the properties have valid values
         if (options.fadeIn < 0) {
            $.error("The fadeIn property of jQuery.auderoFlashingText can't be a negative number");
            return;
         }
         if (options.duration < 0) {
            $.error("The duration property of jQuery.auderoFlashingText can't be a negative number");
            return;
         }
         if (options.fadeOut < 0) {
            $.error("The fadeOut property of jQuery.auderoFlashingText can't be a negative number");
            return;
         }
         if ($.inArray(options.selection, ["random", "ascending", "descending"]) < 0) {
            $.error("The selection property of jQuery.auderoFlashingText should have one of these values: random, ascending, descending");
            return;
         }
         if (options.repeat <= 0 && options.repeat !== -1) {
            $.error("jQuery.auderoFlashingText should run at least one time");
            return;
         }
         if (options.pause !== "random" && options.pause < 0) {
            $.error("jQuery.auderoFlashingText should run the animation with a pause equal to or greater than zero");
            return;
         }
         if (options.speed <= 0) {
            $.error("jQuery.auderoFlashingText should run the animation with a speed greater than zero");
            return;
         }
         if (options.fontMinSize <= 0) {
            $.error("The fontMinSize property of jQuery.auderoFlashingText can't be a negative number");
            return;
         }
         if (options.fontMaxSize <= 0) {
            $.error("The fontMaxSize property of jQuery.auderoFlashingText can't be a negative number");
            return;
         }

         this.css("position", "relative");
         this
            .children()
            .css("visibility", "hidden");
         var $current;
         for (var i = 0; i < this.length; i++) {
            $current = $(this[i]);
            if(options.strings === null) {
               options.strings = $current.children().map(function() {
                  return $(this).text();
               });
            }
            // Clone the options object so elements will have the same values without sharing the same object
            $current.data(dataAttributeName, $.extend(true, {}, options));
            if (options.isEnabled === true) {
               animateText($current, index);
            }
         }

         return this;
      },
      enable: function () {
         var $current;
         for (var i = 0; i < this.length; i++) {
            $current = $(this[i]);
            if (!isRunning($current)) {
               $current.data(dataAttributeName).isEnabled = true;
               animateText($current, $current.data(dataAttributeName).lastIndex);
            }
         }

         return this;
      },
      disable: function () {
         var $current;
         for (var i = 0; i < this.length; i++) {
            $current = $(this[i]);
            if (isRunning($current)) {
               $current.data(dataAttributeName).isEnabled = false;
               $current.stop(true, true);
            }
         }

         return this;
      },
      toggle: function () {
         var $current;
         for (var i = 0; i < this.length; i++) {
            $current = $(this[i]);
            if (isRunning($current)) {
               methods.disable.apply([$current]);
            } else {
               methods.enable.apply([$current]);
            }
         }

         return this;
      },
      destroy: function () {
         for (var i = 0; i < this.length; i++) {
            var $current = $(this[i]);
            if (isRunning($current)) {
               // Removes the floating text
               $current
                  .children(":animated")
                  .stop(true)
                  .fadeOut($current.data(dataAttributeName).fadeOut)
                  .remove();
            }
            $current
               .css("position", "inherit")
               .removeData(dataAttributeName);
            // Restore the default visibility
            this
               .children()
               .css("visibility", "visible");
         }

         return this;
      }
   };

   $.fn.auderoFlashingText = function (method) {
      if (methods[method]) {
         return methods[method].apply(this, Array.prototype.slice.call(arguments, 1));
      } else if (typeof method === "object" || !method) {
         return methods.init.apply(this, arguments);
      } else {
         $.error("Method " + method + " does not exist on jQuery.auderoFlashingText");
      }
   };

   $.fn.auderoFlashingText.defaults = {
      strings: null,       // array (optional). A set of strings to show.
      isEnabled: true,     // boolean (optional). The state of the animation. If the value is false, the animation will stop keeping its current options.
      fadeIn: 300,         // number (optional). The time in milliseconds the element appears by fading to opaque.
      duration: 500,       // number (optional). The time of milliseconds the element stays visible.
      fadeOut: 300,        // number (optional). The time of milliseconds the element disappears by fading to transparent.
      selection: "random", // string (optional). The order of selection of the strings. The possible values are: "random", "ascending", "descending".
      repeat: -1,          // number (optional). The number of times to repeat the animation. -1 means unlimited
      pause: 0,            // number (optional). The time (milliseconds) between animations. Set to "random" to have a random time
      fontMinSize: 7,      // number (optional). The minimum font's size of the strings
      fontMaxSize: 28,     // number (optional). The maximum font's size of the strings
      fontUnit: "px"       // string (optional). The unit to use for the font's size. The value accepted are all those accepted by the browsers, for example: "px", "em", "rem" and so on.
   };
})(jQuery);